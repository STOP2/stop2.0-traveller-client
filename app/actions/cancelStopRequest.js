import config from '../config/config'

export const SEND_STOPREQUEST_FROM_CANCELLATION = 'SEND_STOPREQUEST_CANCELLATION'
export const STOPREQUEST_FROM_CANCELLED = 'STOPREQUEST_CANCELLED'
export const STOPREQUEST_FROM_CANCELLATION_ERROR ='STOPREQUEST_CANCELLATION_ERROR'

export const SEND_STOPREQUEST_DESTINATION_CANCELLATION = 'SEND_STOPREQUEST_CANCELLATION'
export const STOPREQUEST_DESTINATION_CANCELLED = 'STOPREQUEST_CANCELLED'
export const STOPREQUEST_DESTINATION_CANCELLATION_ERROR ='STOPREQUEST_CANCELLATION_ERROR'

const API_ENDPOINT = '/stoprequests/cancel'

export let requestStopRequestCancellation = function(fromVehicle)
{
    return {
        type: fromVehicle ? SEND_STOPREQUEST_FROM_CANCELLATION : SEND_STOPREQUEST_DESTINATION_CANCELLATION,
        stopRequestCancelled: false,
        isCancellingStopRequest: true
    }
}

export let receiveStopRequestCancellationConfirmation = function(fromVehicle)
{
    return {
        type: fromVehicle ? STOPREQUEST_FROM_CANCELLED : STOPREQUEST_DESTINATION_CANCELLED,
        stopRequestCancelled: true,
        isCancellingStopRequest: false
    }
}

export let stopRequestCancellationError = function(fromVehicle, error)
{
    return {
        type: fromVehicle ? STOPREQUEST_FROM_CANCELLATION_ERROR : STOPREQUEST_DESTINATION_CANCELLATION_ERROR,
        stopRequestCancelled: false,
        isCancellingStopRequest: false,
        stopRequestCancellationError: error
    }
}

export let cancelStopRequest = function(requestId, fromVehicle)
{
    return dispatch =>
    {
        dispatch(requestStopRequestCancellation())

        return fetch(config.API_URL + API_ENDPOINT + '?request_id=' + requestId, {
            method: 'POST',
            headers:
            {
                'Accept': 'application/json',
                'Content-Type': 'application/json'
            }
        })
        .then(response => {
            if (response.ok) {
                dispatch(receiveStopRequestCancellationConfirmation(fromVehicle))

                // return false as there are no errors
                return false
            }
            else {
                dispatch(stopRequestCancellationError(fromVehicle))

                // return true as there is an error
                return true
            }
        })
        .catch(error => {
            dispatch(stopRequestCancellationError(error, fromVehicle))

            // return true as there is an error
            return true
        })
    }
}
