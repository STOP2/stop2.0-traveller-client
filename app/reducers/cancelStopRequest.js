import { STOPREQUEST_FROM_CANCELLED, STOPREQUEST_DESTINATION_CANCELLED, SEND_STOPREQUEST_FROM_CANCELLATION, SEND_STOPREQUEST_DESTINATION_CANCELLATION, STOPREQUEST_FROM_CANCELLATION_ERROR, STOPREQUEST_DESTINATION_CANCELLATION_ERROR } from '../actions/cancelStopRequest';
import { RESET_STATE } from '../actions/resetStateAction';
export const initialState = {
  stopRequestFromCancelled: false,
  stopRequestDestinationCancelled: false,
  isCancellingFromStopRequest: false,
  isCancellingDestinationStopRequest: false,
  stopRequestFromCancellationError: null,
  stopRequestDestinationCancellationError: null,
};

const cancelStopRequest = (state = initialState, action) => {
  switch (action.type) {
    case SEND_STOPREQUEST_FROM_CANCELLATION:
      return Object.assign({}, state,
        {
          stopRequestFromCancelled: action.stopRequestCancelled,
          isCancellingFromStopRequest: action.isCancellingStopRequest,
        },
            );

    case STOPREQUEST_FROM_CANCELLED:
      return Object.assign({}, state,
        {
          stopRequestFromCancelled: action.stopRequestCancelled,
          isCancellingFromStopRequest: action.isCancellingStopRequest,
        },
            );

    case STOPREQUEST_FROM_CANCELLATION_ERROR:
      return Object.assign({}, state,
        {
          stopRequestFromCancelled: action.stopRequestCancelled,
          isCancellingFromStopRequest: action.isCancellingStopRequest,
          stopRequestFromCancellationError: action.stopRequestCancellationError,
        },
            );

    case SEND_STOPREQUEST_DESTINATION_CANCELLATION:
      return Object.assign({}, state,
        {
          stopRequestDestinationCancelled: action.stopRequestCancelled,
          isCancellingDestinationStopRequest: action.isCancellingStopRequest,
        },
            );

    case STOPREQUEST_DESTINATION_CANCELLED:
      return Object.assign({}, state,
        {
          stopRequestDestinationCancelled: action.stopRequestCancelled,
          isCancellingDestinationStopRequest: action.isCancellingStopRequest,
        },
            );

    case STOPREQUEST_DESTINATION_CANCELLATION_ERROR:
      return Object.assign({}, state,
        {
          stopRequestDestinationCancelled: action.stopRequestCancelled,
          isCancellingDestinationStopRequest: action.isCancellingStopRequest,
          stopRequestDestinationCancellationError: action.stopRequestCancellationError,
        },
            );

    case RESET_STATE:
      return Object.assign({}, state, initialState);

    default:
      return state;
  }
};

export default cancelStopRequest;
