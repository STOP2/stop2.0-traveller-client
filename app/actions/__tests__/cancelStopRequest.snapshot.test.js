import mockStore from 'redux-mock-store';

import { cancelStopRequest, stopRequestCancellationError } from '../cancelStopRequest';

const store = mockStore();

beforeEach(() => {
  store.clearActions();
});

describe('cancelStopRequest actions', () => {
  it('should handle STOPREQUEST_FROM_CANCELLATION_ERROR action', () => {
    expect(stopRequestCancellationError(true, true)).toMatchSnapshot();
  });

  it('should handle STOPREQUEST_DESTINATION_CANCELLATION_ERROR action', () => {
    expect(stopRequestCancellationError(false, true)).toMatchSnapshot();
  });

  it('should handle SEND_STOPREQUEST_FROM_CANCELLATION action', async () => {
    const response = '';

    fetch.mockResponseSuccess(response);

    await store.dispatch(cancelStopRequest(123, true));

    expect(store.getActions()).toMatchSnapshot();
  });

  it('should handle SEND_STOPREQUEST_DESTINATION_CANCELLATION action', async () => {
    const response = '';

    fetch.mockResponseSuccess(response);

    await store.dispatch(cancelStopRequest(123, false));

    expect(store.getActions()).toMatchSnapshot();
  });
});
