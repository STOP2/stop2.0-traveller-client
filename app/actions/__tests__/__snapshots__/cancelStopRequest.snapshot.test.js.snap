exports[`cancelStopRequest actions should handle SEND_STOPREQUEST_DESTINATION_CANCELLATION action 1`] = `
Array [
  Object {
    "isCancellingStopRequest": true,
    "stopRequestCancelled": false,
    "type": "SEND_STOPREQUEST_DESTINATION_CANCELLATION",
  },
  Object {
    "isCancellingStopRequest": false,
    "stopRequestCancelled": true,
    "type": "STOPREQUEST_DESTINATION_CANCELLED",
  },
]
`;

exports[`cancelStopRequest actions should handle SEND_STOPREQUEST_FROM_CANCELLATION action 1`] = `
Array [
  Object {
    "isCancellingStopRequest": true,
    "stopRequestCancelled": false,
    "type": "SEND_STOPREQUEST_FROM_CANCELLATION",
  },
  Object {
    "isCancellingStopRequest": false,
    "stopRequestCancelled": true,
    "type": "STOPREQUEST_FROM_CANCELLED",
  },
]
`;

exports[`cancelStopRequest actions should handle STOPREQUEST_DESTINATION_CANCELLATION_ERROR action 1`] = `
Object {
  "isCancellingStopRequest": false,
  "stopRequestCancellationError": true,
  "stopRequestCancelled": false,
  "type": "STOPREQUEST_DESTINATION_CANCELLATION_ERROR",
}
`;

exports[`cancelStopRequest actions should handle STOPREQUEST_FROM_CANCELLATION_ERROR action 1`] = `
Object {
  "isCancellingStopRequest": false,
  "stopRequestCancellationError": true,
  "stopRequestCancelled": false,
  "type": "STOPREQUEST_FROM_CANCELLATION_ERROR",
}
`;
